datalad_next/processors/lines_processor.py

```diff
@@ -0,0 +1,123 @@
+""" Generator the emits only complete lines """
+
+from __future__ import annotations
+
+from typing import (
+    Generator,
+    Iterable,
+)
+
+
+__all__ = ['lines_processor']
+
+
+def lines_processor(iterable: Iterable[bytes | str],
```

Comment:
I wonder what convinced you to reject the `finditer` based implementation from https://github.com/datalad/datalad-next/issues/537#issuecomment-1815895654?

I found this to be rather fast. Did you benchmark it? That implementation would also allow for a `keep_end` feature without an implied copy.

Reply:
I made a test, this implementation is about twice as fast

Reply:
I do not see this large speed-up:

```
❯ multitime -n10 python ~/hacking/datalad/next/bm_split_finditer.py
===> multitime results
1: python /home/mih/hacking/datalad/next/bm_split_finditer.py
            Mean        Std.Dev.    Min         Median      Max
real        0.215       0.021       0.196       0.209       0.268       
user        0.182       0.018       0.153       0.185       0.216       
sys         0.036       0.012       0.022       0.037       0.054       

❯ multitime -n10 python ~/hacking/datalad/next/bm_split_linesproc.py
===> multitime results
1: python /home/mih/hacking/datalad/next/bm_split_linesproc.py
            Mean        Std.Dev.    Min         Median      Max
real        0.212       0.018       0.192       0.210       0.244       
user        0.179       0.014       0.162       0.177       0.215       
sys         0.037       0.013       0.018       0.034       0.059       
```

Mean difference for me is below a standard deviation.
[bm_split_linesproc.py.txt](https://github.com/datalad/datalad-next/files/13436044/bm_split_linesproc.py.txt)
[bm_split_finditer.py.txt](https://github.com/datalad/datalad-next/files/13436045/bm_split_finditer.py.txt)

Reply:
I made a small modification that seems to be slightly faster:

```
❯ multitime -n10 python ~/hacking/datalad/next/bm_split_finditer.py
===> multitime results
1: python /home/mih/hacking/datalad/next/bm_split_finditer.py
            Mean        Std.Dev.    Min         Median      Max
real        0.205       0.011       0.197       0.199       0.227       
user        0.179       0.008       0.166       0.176       0.194       
sys         0.030       0.007       0.021       0.029       0.042       
```

[bm_split_finditer.py.txt](https://github.com/datalad/datalad-next/files/13437239/bm_split_finditer.py.txt)

Reply:
I had the hunch that the interpreter/import overhead is disproportionally larger that the implementation performance differences. So I changes the brenchmark scripts to contain an importable function that I could more easily benchmark after all imports are made. This should give a more relevant metric:

This is now much closer to the "about twice" assessment.

```
In [3]: %timeit run_iter()
13.4 ms ± 65.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [4]: %timeit run_linesproc()
8.26 ms ± 87.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```